drivers/at25xxx/at25xxx.c

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright (C) 2019 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_at25xxx
+ * @{
+ *
+ * @file
+ * @brief       Driver for the AT25xxx family of SPI-EEPROMs.
+ *              This also includes M95xxx, 25AAxxx, 25LCxxx,
+ *              CAT25xxx & BR25Sxxx.
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ * @}
+ */
+
+#include <errno.h>
+#include <string.h>
+
+#include "at25xxx.h"
+#include "at25xxx_params.h"
+#include "byteorder.h"
+
+#ifdef USEMODULE_XTIMER
+#include "xtimer.h"
+#define POLL_DELAY_US   (1000)
+#endif
+
+#ifndef min
+#define min(a, b) ((a) > (b) ? (b) : (a))
+#endif
+
+#define PAGE_SIZE   (AT25XXX_PARAM_PAGE_SIZE)
+#define ADDR_LEN    (AT25XXX_PARAM_ADDR_LEN)
+#define ADDR_MSK    ((1UL << ADDR_LEN) - 1)
+
+#define CMD_WREN    (0x6)   /* Write Enable             */
+#define CMD_WRDI    (0x4)   /* Write Disable            */
+#define CMD_RDSR    (0x5)   /* Read Status Register     */
+#define CMD_WRSR    (0x1)   /* Write Status Register    */
+#define CMD_READ    (0x3)   /* Read from Memory Array   */
+#define CMD_WRITE   (0x2)   /* Write to Memory Array    */
+
+#define SR_WIP      (0x01)  /* Write In Progress        */
+#define SR_WEL      (0x02)  /* Write Enable Latch       */
+#define SR_BP0      (0x04)  /* Block Protect 0          */
+#define SR_BP1      (0x08)  /* Block Protect 1          */
+#define SR_SRWD     (0x80)  /* Status Register Write Disable */
+
+static inline void getbus(const at25xxx_t *dev)
+{
+    spi_acquire(dev->params.spi, dev->params.cs_pin, SPI_MODE_0, dev->params.spi_clk);
+}
+
+static inline uint32_t _pos(uint8_t cmd, uint32_t pos)
+{
+    /* first byte is CMD, then addr with MSB first */
+    pos = htonl((pos & ADDR_MSK) | ((uint32_t)cmd << ADDR_LEN));
+    pos >>= 8 * sizeof(pos) - (ADDR_LEN + 8);
+    return pos;
+}
+
+static inline bool _write_in_progress(const at25xxx_t *dev)
+{
+    return spi_transfer_reg(dev->params.spi, dev->params.cs_pin, CMD_RDSR, 0) & SR_WIP;
+}
+
+static inline bool _write_enabled(const at25xxx_t *dev)
+{
+    return spi_transfer_reg(dev->params.spi, dev->params.cs_pin, CMD_RDSR, 0) & SR_WEL;
+}
+
+static size_t _write_page(const at25xxx_t *dev, uint32_t pos, const void *data, size_t len)
+{
+    len = min(len, PAGE_SIZE - (pos & (PAGE_SIZE - 1)));
+    pos = _pos(CMD_WRITE, pos);
+
+    /* wait for previous write to finish - may take up to 5 ms */
+    while (_write_in_progress(dev)) {
+#ifdef USEMODULE_XTIMER
+        xtimer_usleep(POLL_DELAY_US);
```

Comment:
Sadly I cannot provide any numbers, but I think releasing the SPI bus before the delay and re-acquiring it afterwards will result in lower power consumption, as the MCU can go to low power mode while waiting 1ms. Also: If e.g. network devices are hooked up on the same SPI bus, it would have a bad impact on performance to block the bus for up to 5 ms. (E.g. on the `cc110x` we have a round trip time of 3ms, so up to 5ms additional delay could potentially more than double the RTT. Even worse: If the `cc110x` driver does not drain the FIFO fast enough while receiving, a frame gets lost. Other network device drivers have to my limited knowledge less rigid timing requirements, but will likely still benefit from the SPI bus not be blocked for up to 5ms.)

Reply:
Ok, added `spi_release()`/`getbus()`.